lms/djangoapps/completion/models.py

```diff
@@ -73,37 +136,101 @@ def submit_completion(self, user, course_key, block_key, completion):
+              its completion changed.
+            * aggregation_name (string): The name of the aggregated blocks.
+              This is set by the level that the aggregation
+              is occurring. Possible values are "course", "chapter", "sequential", "vertical"
+            * earned (float): The positive sum of the fractional completions of all descendant blocks up
+              to the value of possible.
+            * possible (float): The total sum of the possible completion values of all descendant
+              blocks that are visible to the user. This should be a positive integer.
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the
+            object was newly created by this call.
+
+        Raises:
+            TypeError:
+                If the wrong type is passed for the parameters.
+
+            ValueError:
+                If the value of earned is greater than possible.
+
+            django.core.exceptions.ValidationError:
+                If earned / possible results in a number that is less than 0 or greater than 1 or any float is
+                less than zero.
+
+            RuntimeError:
+                If waffle.ENABLE_COMPLETION_TRACKING is not enabled.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+        self.validate(user, course_key, block_key)
+        if earned > possible:
+            raise ValueError(_('Earned cannot be greater than the possible value.'))
+        if possible > 0.0:
+            percent = earned / possible
         else:
-            # If the feature is not enabled, this method should not be called.  Error out with a RuntimeError.
-            raise RuntimeError(
-                "BlockCompletion.objects.submit_completion should not be called when the feature is disabled."
-            )
-        return obj, isnew
+            percent = 0.0
+
+        obj, is_new = self.update_or_create(
+            user=user,
+            course_key=course_key,
+            aggregation_name=aggregation_name,
+            block_key=block_key,
+            defaults={
+                'percent': percent,
+                'possible': possible,
+                'earned': earned,
+            },
+        )
+        return obj, is_new
+
+    @staticmethod
+    def pre_save(sender, instance, **kwargs):
+        """
+        Validate all fields before saving to database.
+        """
+        instance.full_clean()
```

Comment:
This `pre_save` method doesn't need to be duplicated in `AggregateCompletionManager`, it already exists in the base class `CompletionManager`.  Please remove it from here.